upstream/upstream.go

```diff
@@ -654,11 +666,91 @@ func (u *Upstream) shouldSkip(req *common.NormalizedRequest) (reason error, skip
 		}
 	}
 
-	// TODO evm: if block can be determined from request and upstream is only full-node and block is historical skip
+	// if block can be determined from request and upstream is only full-node and block is historical skip
+	bn, ebn := req.EvmBlockNumber()
+	if ebn != nil || bn <= 0 {
+		return fmt.Errorf("failed to get valid request block number: %w", ebn), false
+	}
+
+	nodeType := u.config.Evm.NodeType
+	if nodeType == "" {
+		detectedNodeType, err := u.detectNodeType()
+		if err != nil {
+			return fmt.Errorf("failed to detect node type: %w", err), false
+		}
+		nodeType = common.EvmNodeType(detectedNodeType)
+	}
+
+	if nodeType == common.EvmNodeTypeFull {
+		lb := req.Network().EvmStatePollerOf(u.Config().Id).LatestBlock()
+
+		if bn < lb-u.config.Evm.MaxAvailableRecentBlocks {
+			return fmt.Errorf("block number %d is out of range (must be >= %d)", bn, lb-u.config.Evm.MaxAvailableRecentBlocks), true
+		}
+	}
 
 	return nil, false
 }
 
+func (u *Upstream) detectNodeType() (common.EvmNodeType, error) {
+	// Step 1: Get the latest block number using the state poller
+	req := &common.NormalizedRequest{}
+	lb := req.Network().EvmStatePollerOf(u.Config().Id).LatestBlock()
+
+	// Block heights to check
+	blockHeights := []int64{
+		1,
+		lb / 2,
+		lb - 1024,
+		lb - 128,
+		lb - 32,
+	}
+
+	// Track responses for each block height
+	responses := make(map[int64]bool)
+
+	for _, blockHeight := range blockHeights {
+		// Normalize the block height to hex format
+		hexBlockHeight, err := common.NormalizeHex(fmt.Sprintf("%x", blockHeight))
+
+		if err != nil {
+			return "", err
+		}
+
+		balanceReq := common.NewNormalizedRequest([]byte(fmt.Sprintf(
+			`{"jsonrpc":"2.0","id": %d,"method":"eth_getBlockByNumber","params":["%s", false]}`,
+			util.RandomID(),
+			hexBlockHeight,
+		)))
+
+		_, errBlock := u.Forward(context.Background(), balanceReq)
+		if errBlock == nil {
+			// No error, store response for this block height
+			responses[blockHeight] = true
```

Comment:
Probably it's better to also check if result exists and e.g. has result[hash] of that block, sometimes nodes might give no error but result is nil